internal/app/machined/pkg/controllers/config/acquire.go

```diff
@@ -0,0 +1,437 @@
+	Name() string
+	Configuration(context.Context) ([]byte, error)
+}
+
+// PlatformEventer sends events based on the config process via platform-specific interface.
+type PlatformEventer interface {
+	FireEvent(context.Context, platform.Event)
+}
+
+// Setter sets the current machine config.
+type Setter interface {
+	SetConfig(config.Provider) error
+}
+
+// AcquireController loads the machine configuration from multiple sources.
+type AcquireController struct {
+	PlatformConfiguration PlatformConfigurator
+	PlatformEvent         PlatformEventer
+	ConfigSetter          Setter
+	EventPublisher        talosruntime.Publisher
+	ValidationMode        validation.RuntimeMode
+	ConfigPath            string
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *AcquireController) Name() string {
+	return "config.AcquireController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *AcquireController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: v1alpha1.NamespaceName,
+			Type:      v1alpha1.AcquireConfigSpecType,
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: configresource.NamespaceName,
+			Type:      configresource.MachineConfigType,
+			ID:        pointer.To(configresource.MaintenanceID),
+			Kind:      controller.InputWeak,
+		},
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.MaintenanceServiceRequestType,
+			Kind:      controller.InputDestroyReady,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *AcquireController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: v1alpha1.AcquireConfigStatusType,
+			Kind: controller.OutputExclusive,
+		},
+		{
+			Type: runtime.MaintenanceServiceRequestType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// stateMachineFunc represents the state machine of config.AcquireController.
+type stateMachineFunc func(context.Context, controller.Runtime, *zap.Logger) (stateMachineFunc, config.Provider, error)
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *AcquireController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	if ctrl.ConfigPath == "" {
+		ctrl.ConfigPath = constants.ConfigPath
+	}
+
+	// start always with loading config from disk
+	var currentState stateMachineFunc = ctrl.stateDisk
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		// check the spec first
+		_, err := safe.ReaderGet[*v1alpha1.AcquireConfigSpec](ctx, r, v1alpha1.NewAcquireConfigSpec().Metadata())
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// spec is not found, wait for it
+				continue
+			}
+
+			return fmt.Errorf("failed to get spec: %w", err)
+		}
+
+		// run the state machine
+		for {
+			newState, cfg, err := currentState(ctx, r, logger)
+			if err != nil {
+				ctrl.EventPublisher.Publish(ctx, &machineapi.ConfigLoadErrorEvent{
```

Comment:
what's different between EventPublisher and PlatformEvent?

Reply:
this is part of extracting from v1alpha1_sequencer:

* `EventPublisher` is Talos event stream
* `PlatformEvent` is an event sent via platform-specific API, only implemented for Equinix Metal